symengine/solve.cpp

```diff
@@ -335,4 +337,70 @@ RCP<const Set> solve(const RCP<const Basic> &f, const RCP<const Symbol> &sym,
     // TODO - Trig solver
     return solve_rational(newf, sym, domain);
 }
+
+vec_basic linsolve_helper(const DenseMatrix &A, const DenseMatrix &b)
+{
+    DenseMatrix res(A.nrows(), 1);
+    fraction_free_gauss_jordan_solve(A, b, res);
+    vec_basic fs;
+    for (unsigned i = 0; i < res.nrows(); i++) {
+        fs.push_back(res.get(i, 0));
+    }
+    return fs;
+}
+
+vec_basic linsolve(const DenseMatrix &system, const vec_sym &syms)
+{
+    DenseMatrix A(system.nrows(), system.ncols() - 1), b(system.nrows(), 1);
+    system.submatrix(A, 0, 0, system.nrows() - 1, system.ncols() - 2);
+    system.submatrix(b, 0, system.ncols() - 1, system.nrows() - 1,
+                     system.ncols() - 1);
+    return linsolve_helper(A, b);
+}
+
+vec_basic linsolve(const vec_basic &system, const vec_sym &syms)
+{
+    DenseMatrix A, b;
+    auto mat = linear_eqns_to_matrix(system, syms);
+    A = mat.first;
+    b = mat.second;
+    return linsolve_helper(A, b);
+}
+
+std::pair<const DenseMatrix, const DenseMatrix>
+linear_eqns_to_matrix(const vec_basic &equations, const vec_sym &syms)
+{
+    auto size = numeric_cast<unsigned int>(syms.size());
+    DenseMatrix A(1, size);
+    vec_basic coeffs, bvec;
+    RCP<const Basic> tmp;
+    RCP<const Number> coef;
+    umap_basic_num d;
+
+    int row = 0;
+    for (const auto &eqn : equations) {
+        coeffs.clear();
+        d.clear();
+        coef = zero;
+
+        auto neqn = eqn;
+        if (is_a<Equality>(*eqn)) {
+            neqn = sub(down_cast<const Equality &>(*eqn).get_arg2(),
+                       down_cast<const Equality &>(*eqn).get_arg1());
+        }
+        for (const auto &sym : syms) {
+            auto nc = coeff(*neqn, *sym, *one);
```

Comment:
This traverses the expression tree for each of the iteration? Is there a better way?

Reply:
Which equation is different in every iteration?

Reply:
@ranjithkumar007, can you answer @isuruf's question here please?

Reply:
@isuruf , Although I am not 100% sure, I think one possible way to avoid traversing the expression tree for every iteration would be modifying the `coeff` visitor(new derived visitor or something similar) to accommodate a vector of symbols.

Reply:
Or you could convert it to a poly and get the coeffs.

Reply:
@certik, I would like to do what @isuruf suggested here before this gets in.

Reply:
Thanks for the suggestion @isuruf